tests/end-to-end/upgrade/internal/runner/runner.go

```diff
@@ -0,0 +1,247 @@
+package runner
+
+import (
+	"bytes"
+	"fmt"
+	"regexp"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/kyma-project/kyma/tests/end-to-end/upgrade/internal/platform/logger"
+
+	"github.com/pkg/errors"
+	uuid "github.com/satori/go.uuid"
+	"github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	gracefulTimeout = time.Second * 10
+	runnerLabelName = "upgrade.tester.kyma-project.io"
+	// negative value for regex used for name validation in k8s:
+	// https://github.com/kubernetes/apimachinery/blob/98853ca904e81a25e2000cae7f077dc30f81b85f/pkg/util/validation/validation.go#L110
+	regexSanitize = "[^a-z0-9]([^-a-z0-9]*[^a-z0-9])?"
+)
+
+// NamespaceCreator has methods required to create k8s ns.
+type NamespaceCreator interface {
+	Create(*v1.Namespace) (*v1.Namespace, error)
+}
+
+// TestRunner executes registered tests
+type TestRunner struct {
+	log                 *logrus.Entry
+	nsCreator           NamespaceCreator
+	tests               map[string]UpgradeTest
+	maxConcurrencyLevel int
+	sanitizeRegex       *regexp.Regexp
+}
+
+// NewTestRunner is a constructor for TestRunner
+func NewTestRunner(log *logrus.Entry, nsCreator NamespaceCreator, tests map[string]UpgradeTest, maxConcurrencyLevel int) (*TestRunner, error) {
+	sanitizeRegex, err := regexp.Compile(regexSanitize)
+	if err != nil {
+		return nil, errors.Wrap(err, "while compiling sanitize regexp")
+	}
+
+	return &TestRunner{
+		log:                 log.WithField("service", "test:runner"),
+		nsCreator:           nsCreator,
+		tests:               tests,
+		maxConcurrencyLevel: maxConcurrencyLevel,
+		sanitizeRegex:       sanitizeRegex,
+	}, nil
+}
+
+// PrepareData executes CreateResources method for each registered test.
+// Test are run in parallel with given maxConcurrencyLevel
+func (r *TestRunner) PrepareData(stopCh <-chan struct{}) error {
+	wg := sync.WaitGroup{}
+	queue := make(chan task)
+
+	failedTaskCnt := 0
+	// spawn workers
+	for i := 0; i < r.maxConcurrencyLevel; i++ {
+		wg.Add(1)
+		go func() {
+			for task := range queue {
+				failed := r.executeTask(task.CreateResources, stopCh, "CreateResources", task.Name(), true)
+				if failed {
+					failedTaskCnt++
+				}
+			}
+			wg.Done()
+		}()
+	}
+
+	// populate all tests
+	for name, test := range r.tests {
+		queue <- task{name, test}
+	}
+	close(queue)
+
+	// wait for the workers to finish
+	r.wgWait(stopCh, &wg)
+
+	if failedTaskCnt > 0 {
+		return fmt.Errorf("executed %d task and %d of them failed", len(r.tests), failedTaskCnt)
+	}
+	return nil
+}
+
+// ExecuteTests executes TestResources method for each registered test.
+// Test are run in parallel with given maxConcurrencyLevel
+func (r *TestRunner) ExecuteTests(stopCh <-chan struct{}) error {
```

Comment:
`ExecuteTests` and `PrepareData` are almost the same. Have you considered make one method with extra argument `CreateResources/TestResources`? For quick search i found this ugly solution :) to run `task` struct method:
```
m := reflect.ValueOf(struct_type).MethodByName("func_name")
m.Call(nil)
```
maybe there is something prettier, and we do not have to repeat the code or maybe it is on purpose because method will change in future?

Reply:
it's not so easy to unify that. I already did that with the `executeTaskFunc` method. With your solution, u didnt take into account the channels and closure in the method. 

Please provide e2e proposal then we can check if it will be easier to maintain and understand, right now I do not see that option